WT-6220 Add statistics tracking when times are written to disk

agorrod opened the pull request.

agorrod commented:
@luke-pearson and @keithbostic would you mind taking a look at this change, please? It is attempting to output statistics related to how much time information is output to disk. Feedback is welcome on both the set of statistics added and how I'm tracking and incrementing the stats.

luke-pearson commented on src/include/reconcile.h:
```diff
@@ -143,6 +143,27 @@ struct __wt_reconcile {
+     * is written to disk. The number of entries on a page is limited to a 32 bit number so these
+     * counters can be too.
+     */
+    uint32_t count_durable_start_ts;
```
Is it worth moving these together with the stats added in  WT-6080? i.e.  bool rec_page_cell_with_ts;

agorrod replied:
Ah - yup. There is a lot of overlap. I'll touch base with Haseeb. Thanks for the heads up.

hbokhari commented on dist/stat_data.py:
```diff
@@ -800,6 +821,34 @@ def __init__(self, name, desc, flags=''):
     CursorStat('cursor_update_bytes_changed', 'update value size change', 'size'),
 
+    ##########################################
+    # History statistics
```
@agorrod, I am curious about prefix hs_*. My first impression when reading through the files was that these have something to do with history store (hs) only realizing afterwards that I made an incorrect assumption.

agorrod replied:
Sure. I chose hs - because I think of timestamps and transaction IDs as forming the history, so it seemed natural to put these in the History statistics section and `hs` was the common prefix there. I'm not wedded to the prefix - feel free to suggest something different.

keithbostic replied:
+1.

hbokhari replied:
I would personally like them to be under reconciliation stats. Perhaps something like this will make be appropriate:

HistoryStat('rec_internal_pages_with_start_durable_ts', 'Internal pages written that include a newest start durable timestamp '),
HistoryStat('rec_leaf_pages_with_start_durable_ts', 'leaf pages written that included at least one start durable timestamp),
HistoryStat('rec_records_with_start_durable_ts', 'records written that included a start durable timestamp'),

Note that this statistics have a wider audience (e.g., TSEs) and it maybe easier to explain "internal" pages vs. "address" pages.

keithbostic commented on dist/stat_data.py:
```diff
@@ -397,6 +397,27 @@ def __init__(self, name, desc, flags=''):
+    HistoryStat('hs_write_rec_start_ts', 'records written that included a start timestamp'),
+    HistoryStat('hs_write_rec_start_txn', 'records written that included a start transaction ID'),
+    HistoryStat('hs_write_rec_stop_ts', 'records written that included a stop timestamp'),
+    HistoryStat('hs_write_rec_stop_txn', 'records written that included a stop transaction ID'),
```
The list should be sorted?

keithbostic replied:
"that included" -> "including"?

keithbostic commented on dist/stat_data.py:
```diff
@@ -397,6 +397,27 @@ def __init__(self, name, desc, flags=''):
+    HistoryStat('hs_write_rec_start_ts', 'records written that included a start timestamp'),
+    HistoryStat('hs_write_rec_start_txn', 'records written that included a start transaction ID'),
+    HistoryStat('hs_write_rec_stop_ts', 'records written that included a stop timestamp'),
+    HistoryStat('hs_write_rec_stop_txn', 'records written that included a stop transaction ID'),
```
Is it interesting to tease apart start txn IDs vs. start timestamps? In other code I've not done that, I've just say "a start thing got set", so there are 3 categories, "start", "stop" and "durable". If it's not interesting to split them apart, that simplifies things.

agorrod replied:
I believe it is interesting to have different stats for txnID vs timestamp. Generally I expect that the window of required timestamps can be long seeing issues because of that might point towards suggesting applications change how they use MongoDB.
I'm hoping we pretty much never see transaction IDs written to disk, if we do that's a pointer to somewhere we should look to make a WiredTiger of MongoDB change to avoid.

keithbostic commented on dist/stat_data.py:
```diff
@@ -397,6 +397,27 @@ def __init__(self, name, desc, flags=''):
+    HistoryStat('hs_write_addr_oldest_start_ts', 'page addresses written that include a oldest start timestamp '),
+    HistoryStat('hs_write_addr_oldest_start_txn', 'page addresses written that include a oldest start transaction ID '),
+    HistoryStat('hs_write_addr_newest_stop_ts', 'page addresses written that include a newest stop timestamp '),
+    HistoryStat('hs_write_addr_newest_stop_txn', 'page addresses written that include a newest stop transaction ID'),
```
Not a big fan of "page addresses", maybe "internal page"?

Don't feel strongly.

keithbostic commented on dist/stat_data.py:
```diff
@@ -397,6 +397,27 @@ def __init__(self, name, desc, flags=''):
     HistoryStat('hs_gc_pages_removed', 'history pages removed for garbage collection'),
     HistoryStat('hs_gc_pages_visited', 'history pages visited for garbage collection'),
+    HistoryStat('hs_write_addr_prepared', 'page addresses written that included at least one prepare state'),
+    HistoryStat('hs_write_addr_newest_start_durable_ts', 'page addresses written that include a newest start durable timestamp '),
```
I agree pretty strongly with @hbokhari, "hs" is a history store prefix, and these aren't history store statistics.

Reconciliation, maybe?

keithbostic commented on dist/stat_data.py:
```diff
@@ -800,6 +821,34 @@ def __init__(self, name, desc, flags=''):
+    HistoryStat('hs_write_rec_start_ts', 'records written that included a start timestamp'),
+    HistoryStat('hs_write_rec_start_txn', 'records written that included a start transaction ID'),
+    HistoryStat('hs_write_rec_stop_ts', 'records written that included a stop timestamp'),
+    HistoryStat('hs_write_rec_stop_txn', 'records written that included a stop transaction ID'),
```
This is a lot of new statistics. I was trying to think if internal statistics are interesting here, be interested to hear the use case you have in mind? I can see leaf-page counts, but I'm not sure when I'd want to see internal page counts.

keithbostic commented on src/include/reconcile.h:
```diff
@@ -143,6 +143,27 @@ struct __wt_reconcile {
+    bool has_oldest_start_txn;
+    bool has_newest_stop_ts;
+    bool has_newest_stop_txn;
+    bool has_prepare;
```
The `bool` type is 4B on Windows, so we can shrink this down by using `uint8_t`.

keithbostic commented on src/reconcile/rec_write.c:
```diff
@@ -1676,6 +1676,90 @@ __rec_compression_adjust(WT_SESSION_IMPL *session, uint32_t max, size_t compress
 
+/*
+ * __rec_page_time_stats --
+ *     Update statistics about this cell, don't bother clearing here - they are cleared in a
```
"about this page"?

agorrod commented:
Thanks for the reviews @hbokhari and @keithbostic. I've addressed most of your comments.

There was one outstanding question about the number of statistics being added. I think this is new enough part of the database that I'd rather have more information to help diagnose issues rather than less. If some of these statistics turn out to not be useful in the future I'll be happy to remove them. If you feel strongly - we could possibly remove the aggregated version of the statistics, but I will be interested to see if they ever diverge from the leaf page versions.

agorrod commented:
@keithbostic would you mind taking another look at this failure?

Also: I'm (embarasingly) not able to regenerate wiredtiger.in at the moment - it's changing a bunch of &nbsp; characters in unexpected ways. Would you mind running s_all and pushing the results into the PR (if you think it's worth keeping)?

keithbostic commented:
@agorrod, I'm not sure what you mean by taking another look at the failure? I pushed the `s_all` changes into the ticket, and tests are passing.

agorrod commented:
bq. I'm not sure what you mean by taking another look at the failure?

Sorry - I wasn't concentrating. I meant to ask for another round of review.

Thanks for fixing up s_all..

keithbostic commented on src/include/reconcile.i:
```diff
@@ -116,6 +116,29 @@ __wt_rec_image_copy(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_REC_KV *kv)
+    if (ta->prepare != 0)
+        FLD_SET(r->ts_usage_flags, WT_REC_TIME_PREPARE);
+}
+
```
Maybe it's time for `reconcile/rec_stat.i`? (I don't feel strongly.)

keithbostic commented on src/reconcile/rec_write.c:
```diff
@@ -1676,6 +1676,90 @@ __rec_compression_adjust(WT_SESSION_IMPL *session, uint32_t max, size_t compress
+ * __rec_page_time_stats --
+ *     Update statistics about this page, don't bother clearing here - they are cleared in a
+ *     different place, because sometimes we accumulate these stats and don't end up writing.
+ */
```
This doesn't appear to be correct, the function is only called from one place and could reasonable clear the fields.

agorrod replied:
It's the goto `copy_image` that's important, I think. We want to cleanup regardless of whether we write the page out, but only want to record stats when the page is actually written out.

keithbostic replied:
Got it, understood.

keithbostic commented:
@agorrod, I've pushed the changes I suggested to the base branch in 36797e2 and the additional changes to track the maximum reconciliation time in 629a42a.

If you're happy with both, let's merge.

agorrod commented on src/include/connection.h:
```diff
@@ -312,6 +312,7 @@ struct __wt_connection_impl {
     uint32_t stat_flags; /* Options declared in flags.py */
 
     /* Connection statistics */
+    uint64_t rec_maximum_seconds; /* Maximum seconds reconciliation took. */
```
This doesn't look quite complete to me - I expected to see a WT_STAT_SET and explicit reset call similar to the max page statistic:
```
$ ack evict_max_page_size src/
src/conn/conn_cache.c
300:    WT_STAT_SET(session, stats, cache_eviction_maximum_page_size, cache->evict_max_page_size);
src/evict/evict_page.c
205:    if (page->memory_footprint > conn->cache->evict_max_page_size)
206:        conn->cache->evict_max_page_size = page->memory_footprint;
src/include/cache.h
91:    uint64_t evict_max_page_size; /* Largest page seen at eviction */
src/txn/txn_ckpt.c
777:    conn->cache->evict_max_page_size = 0;
```

keithbostic replied:
Thanks, you're correct -- pushed in 92d61f92e.

I don't see any obvious clean way to do this (maybe it's time for a new class of "reconciliation" statistics, separate from the "cache" statistics?).

Let me know if this still doesn't look correct.

agorrod commented:
evergreen retry

agorrod commented:
@keithbostic and @hbokhari. I'm good to go with this change now - could you give a confirmation that you are as well, since I've managed to share ownership across three people?

keithbostic commented:
@agorrod, I'm good to go.

hbokhari commented:
@agorrod, LGTM
